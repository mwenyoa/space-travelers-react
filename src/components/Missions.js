import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { displayMissions } from '../Redux/missions/missionReducer';

let triggerOnce = false;

const Missions = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (triggerOnce) return;
    dispatch(displayMissions());
    triggerOnce = true;
  }, []);

  return (
    <div>Hello</div>
  );
};
export default Missions;
