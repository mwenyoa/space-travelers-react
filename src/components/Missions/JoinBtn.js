import React from 'react';
import { useDispatch } from 'react-redux';
import { Button } from 'react-bootstrap';
import propTypes from 'prop-types';
import {
  handleJoinButton,
  handleLeaveButton,
} from '../../Redux/missions/missionReducer';

const JoinBtn = ({ id, reserved }) => {
  const dispatch = useDispatch();
  let button;
  if (!reserved) {
    button = (
      <Button
        variant="outline-secondary"
        type="Button"
        className="text-nowrap"
        onClick={() => dispatch(handleJoinButton(id))}
      >
        Join Mission
      </Button>
    );
  } else {
    button = (
      <Button
        variant="outline-danger"
        type="Button"
        className="text-nowrap"
        onClick={() => dispatch(handleLeaveButton(id))}
      >
        Leave Mission
      </Button>
    );
  }
  return (
    <div>
      {button}
    </div>
  );
};

JoinBtn.propTypes = {
  id: propTypes.string.isRequired,
  reserved: propTypes.bool.isRequired,
};

export default JoinBtn;
