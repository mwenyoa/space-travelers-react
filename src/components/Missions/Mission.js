import React from 'react';
import propTypes from 'prop-types';
import { Badge } from 'react-bootstrap';
import JoinBtn from './JoinBtn';

const StatusBadge = ({ reserved }) => {
  let badge;
  if (!reserved) {
    badge = (
      <Badge bg="secondary">
        Not a member
      </Badge>
    );
  } else {
    badge = (
      <Badge bg="success">
        Active member
      </Badge>
    );
  }
  return (
    <div>
      {badge}
    </div>
  );
};

const Mission = ({
  id, name, description, reserved,
}) => (
  <tr key={id}>
    <td><b>{name}</b></td>
    <td>{description}</td>
    <td className="align-middle">
      <StatusBadge reserved={reserved} />
    </td>
    <td className="align-middle">
      <JoinBtn id={id} reserved={reserved} />
    </td>
  </tr>
);

Mission.propTypes = {
  id: propTypes.string.isRequired,
  name: propTypes.string.isRequired,
  description: propTypes.string.isRequired,
  reserved: propTypes.bool.isRequired,
};

StatusBadge.propTypes = {
  reserved: propTypes.bool.isRequired,
};

export default Mission;
