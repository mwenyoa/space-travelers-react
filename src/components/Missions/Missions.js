import React, { useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import 'bootstrap/dist/css/bootstrap.css';
import Table from 'react-bootstrap/Table';
import { displayMissions } from '../../Redux/missions/missionReducer';
import Mission from './Mission';

let triggerOnce = false;

const Missions = () => {
  const missionList = useSelector(
    (state) => state.missionReducer,
    shallowEqual,
  ).flat();

  const dispatch = useDispatch();

  useEffect(() => {
    if (triggerOnce) return;
    dispatch(displayMissions());
    triggerOnce = true;
  }, []);

  return (
    <div>
      <Table
        responsive
        striped
        bordered
        hover
        className="mt-4 mx-auto w-75"
        size="sm"
      >
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          {missionList.map(({ id, name, description }) => (
            <Mission key={id} id={id} name={name} description={description} />
          ))}
        </tbody>
      </Table>
    </div>
  );
};
export default Missions;
