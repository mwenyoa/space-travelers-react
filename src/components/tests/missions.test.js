import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../../Redux/ConfigureStore';
import Missions from '../Missions/Missions';
import Mission from '../Missions/Mission';

it('Mission render correctly', () => {
  const mission = {
    id: '9D1B7E0',
    name: 'Thaicom',
    description:
      'ThaThaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited, which is the company that owns and operates the Thaicom satellite fleet and other telecommunication businesses in Thailand and throughout the Asia-Pacific region. The satellite projects were named Thaicom by the King of Thailand, His Majesty the King Bhumibol Adulyadej, as a symbol of the linkage between Thailand and modern communications technology.',
    reserved: false,
  };

  const tree = renderer.create(
    <Provider store={store}>
      <Mission
        id={mission.id}
        name={mission.name}
        description={mission.description}
        reserved={mission.reserved}
      />
    </Provider>,
  );
  expect(tree).toMatchSnapshot();
});

it('Missions renders correctly', () => {
  const list = renderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>,
  );
  expect(list).toMatchSnapshot();
});
