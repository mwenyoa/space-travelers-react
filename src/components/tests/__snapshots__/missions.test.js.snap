// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission render correctly 1`] = `
<tr>
  <td>
    <b>
      Thaicom
    </b>
  </td>
  <td>
    ThaThaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited, which is the company that owns and operates the Thaicom satellite fleet and other telecommunication businesses in Thailand and throughout the Asia-Pacific region. The satellite projects were named Thaicom by the King of Thailand, His Majesty the King Bhumibol Adulyadej, as a symbol of the linkage between Thailand and modern communications technology.
  </td>
  <td
    className="align-middle"
  >
    <div>
      <span
        className="badge bg-secondary"
      >
        Not a member
      </span>
    </div>
  </td>
  <td
    className="align-middle"
  >
    <div>
      <button
        className="text-nowrap btn btn-outline-secondary"
        disabled={false}
        onClick={[Function]}
        type="Button"
      >
        Join Mission
      </button>
    </div>
  </td>
</tr>
`;

exports[`Missions renders correctly 1`] = `
<div>
  <div
    className="table-responsive"
  >
    <table
      className="mt-4 mx-auto w-75 table table-sm table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
